modules/react/combobox/lib/hooks/useKeyboardTypeAhead.ts

```diff
@@ -0,0 +1,132 @@
+import React from 'react';
+import {createElemPropsHook} from '@workday/canvas-kit-react/common';
+import {useComboboxModel} from './useComboboxModel';
+
+/**
+ * `useKeyboardTypeAhead` handles type-ahead. It will try and find a matching item in the list. It also returns the keys types so far.
+ * This hook is best used when composing hooks and in conjunction with `useComboboxInput`.
+ *
+ * ```tsx
+ * // Example for a `Select` input
+ * export const useSelectInput = composeHooks(
+ * createElemPropsHook(useComboboxModel)((model, ref, elemProps: {keySofar?: string} = {}) => {
+ *  return {
+ *    onKeyDown(event: React.KeyboardEvent) {
+ *     // onKeyDown logic ...
+ *    },
+ *  } as const;
+ * }),
+ * useKeyboardTypeAhead,
+ * useComboboxInput
+ * );
+ * ```
+ */
+export const useKeyboardTypeAhead = createElemPropsHook(useComboboxModel)(model => {
+  const keySofar = React.useRef('');
+  const timer = React.useRef<ReturnType<typeof setTimeout>>();
+  const [keyTypedSofar, setKeyTypedSofar] = React.useState(keySofar.current);
+
+  const startClearKeysSoFarTimer = () => {
+    if (timer.current) {
+      clearTimeout(timer.current);
+    }
+    timer.current = setTimeout(() => {
+      keySofar.current = '';
+      setKeyTypedSofar(keySofar.current);
+    }, 500);
+  };
+
+  // Based on the the key typed, this will try and find an index in the array where it matches the string
+  // so if a user types `de` it will try and find an index in the array whose id matches `de` like denver
+  const getIndexByStartString = (
+    startIndex: number,
+    startString: string,
+    endIndex: number = model.state.items.length,
+    ignoreDisabled: boolean = true
+  ): number => {
+    for (let i = startIndex; i < endIndex; i++) {
+      const label = model.state.items[i].textValue.toLowerCase();
+      if (label.indexOf(startString.toLowerCase()) === 0) {
+        if (
+          !ignoreDisabled ||
+          (ignoreDisabled && !model.state.nonInteractiveIds.includes(model.state.items[i].id))
+        ) {
+          return i;
+        }
+      }
+    }
```

Comment:
```suggestion
      if (
        label.startsWith(startString.toLowerCase()) &&
        (!ignoreDisabled || !model.state.nonInteractiveIds.includes(model.state.items[i].id))
      ) {
        return i;
      }
```
`startsWith` will check if label starts from startString (i guess it is what you check)
`ignoreDisabled` is boolean, so there is only two variations. You already check one of this, so you don't need to check again here:
`ignoreDisabled && !model.state.nonInteractiveIds.includes(model.state.items[i].id))`

Reply:
i updated this with starts with, but the rest of the logic didn't work

Reply:
```if (
        label.startsWith(startString.toLowerCase()) &&
        (!ignoreDisabled ||
          (ignoreDisabled && !model.state.nonInteractiveIds.includes(model.state.items[i].id)))
      ) {
        return i;
      }
      ```

Reply:
@RayRedGoose I think you missed the second conditional. In English "if label starts with `startString` AND either `ignoreDisabled` is `false` OR `ignoreDisabled` is `true` AND the current item is interactable". Your logic is missing the "`ignoreDisabled` is `true`" part.

Here's logic that's the same:
```suggestion
      if (
        label.startsWith(startString.toLowerCase()) &&
        (!ignoreDisabled ||
          (ignoreDisabled && !model.state.nonInteractiveIds.includes(model.state.items[i].id)))
      ) {
        return i;
      }
```